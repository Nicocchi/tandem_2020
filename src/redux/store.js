import { createStore, applyMiddleware, compose } from "redux";
import thunk from "redux-thunk";
import logger from "redux-logger";
import combineReducer from "./reducers";

let store = null;

// Redux dev tools
const composeEnhancers =
    typeof window === "object" && window.__REDUX_DEVTOOLS_EXTENSION_COMPOSE__
        ? window.__REDUX_DEVTOOLS_EXTENSION_COMPOSE__({
              // Specify extensionâ€™s options like name, actionsBlacklist, actionsCreators, serialize...
          })
        : compose;

const devEnhancer = composeEnhancers(applyMiddleware(thunk, logger));
const enhancer = composeEnhancers(applyMiddleware());

// Create redux store
if (process.env.REACT_APP_NODE_ENV !== "production") {
    store = createStore(combineReducer, devEnhancer);
} else {
    store = createStore(combineReducer, enhancer);
}

export default store;