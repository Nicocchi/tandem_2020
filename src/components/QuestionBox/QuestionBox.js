import React, { useState, useEffect } from "react";
import PropTypes from "prop-types";
import { connect } from "react-redux";
import { increaseScore, increaseCorrect, increaseIncorrect } from "../../redux/actions";
import shortid from "shortid";
import { shuffle } from "../../Utils";

function QuestionBox(props) {
    const [isCorrect, setIsCorrect] = useState(false);
    const [isFinished, setIsFinished] = useState(false);
    const [answers, setAnswers] = useState([]);

    useEffect(() => {
        if (props.question?.correct && props.question?.incorrect) {
            setAnswers(shuffle([...props.question.incorrect, props.question.correct]));
        }
        return () => {
            
        }
    }, [props.question])

    const select = (index) => {
        if (isFinished) return;
        if (index === props.question.correct) {
            setIsCorrect(true);
            props.increaseCorrect();
            props.increaseScore();
        } else {
            props.increaseIncorrect();
        }

        setIsFinished(true);
    };

    return (
        <div>
            <p className="question animated fade-in">
                {props.current + 1} / {props.endAmount}
            </p>
            <p className="question animated fade-in" data-testid="question">
                {props.question?.question}
            </p>
            {
                <div style={{ display: isFinished ? "none" : "flex" }} className="answers-box">
                    {answers.map((qst, index) => (
                        <button
                            style={{
                                animationDuration: index === 0 ? `0.2s` : `${index / 2}s`,
                                animationFillMode: "both",
                            }}
                            className={`answer-btn ${!isFinished ? "bounce-in-right" : null}`}
                            key={shortid.generate()}
                            onClick={() => select(qst)}
                        >
                            {qst}
                        </button>
                    ))}
                </div>
            }

            <div className="answer-wrapper">
                {isFinished ? (
                    <button className="answer-btn next-btn animated flip-in-y" onClick={props.nextQuestion}>
                        Next
                    </button>
                ) : null}
                {isFinished && isCorrect ? (
                    <p className="correct-text">Correct!</p>
                ) : isFinished && !isCorrect ? (
                    <span className="answer-text">
                        <p className="incorrect-text">Incorrect! The correct answer is:</p>
                        <p className="correct-text">{props.question.correct}</p>
                    </span>
                ) : null}
            </div>
        </div>
    );
}

QuestionBox.propTypes = {
    question: PropTypes.exact({
        question: PropTypes.string.isRequired,
        correct: PropTypes.string.isRequired,
        incorrect: PropTypes.arrayOf(PropTypes.string).isRequired,
    }).isRequired,
    nextQuestion: PropTypes.func.isRequired,
};

export default connect(null, { increaseScore, increaseCorrect, increaseIncorrect })(QuestionBox);
