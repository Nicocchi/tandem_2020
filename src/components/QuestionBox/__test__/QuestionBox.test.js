import React from "react";
import ReactDOM from "react-dom";
import QuestionBox from "./../QuestionBox";
import store from "../../../redux/store";
import { Provider } from "react-redux";

import { render, cleanup } from "@testing-library/react";
import "@testing-library/jest-dom/extend-expect";

import renderer from "react-test-renderer";
import sinon from "sinon";

const question = {
    correct: "Unicorn", incorrect: ["Bear", "Rabbit", "Seal"], question: "What is the national animal of Scotland?"
}

describe('QuestionBox', () => {
    afterEach(cleanup);

    function mockFunction() {

    }

    it("renders without crashing with props", () => {
        const div = document.createElement('div');
        ReactDOM.render(<Provider store={store}><QuestionBox question={question} nextQuestion={mockFunction} /></Provider>, div);
    });

    it("throws an error when no props are not passed", () => {
        const div = document.createElement('div');
        const stub = sinon.stub(console, 'error');

        ReactDOM.render(<Provider store={store}><QuestionBox /></Provider>, div);
        expect(stub.calledTwice).toEqual(true);
        expect(stub.calledOnceWithExactly('Warning: Failed prop type: The prop `question` is marked as required in `QuestionBox`, but its value is `undefined`.'))
        expect(stub.calledOnceWithExactly('Warning: Failed prop type: The prop `nextQuestion` is marked as required in `QuestionBox`, but its value is `undefined`.'))
        console.error.restore();
    });

    it("throws an error when one prop is not passed", () => {
        const div = document.createElement('div');
        const stub = sinon.stub(console, 'error');

        ReactDOM.render(<Provider store={store}><QuestionBox question={question} /></Provider>, div);
        expect(stub.calledOnceWithExactly('Warning: Failed prop type: The prop `nextQuestion` is marked as required in `QuestionBox`, but its value is `undefined`.'))
        console.error.restore();
    });

    it("matches snapshop", () => {
        const tree = renderer.create(<Provider store={store}><QuestionBox question={question} /></Provider>).toJSON();
        expect(tree).toMatchSnapshot();
    })

    describe('question', () => {
        it('renders the text correctly', () => {
            const div = document.createElement('div');
            const {getByTestId} = render(<Provider store={store}><QuestionBox question={question} nextQuestion={mockFunction} /></Provider>);
            expect(getByTestId('question')).toHaveTextContent(question.question);
        })
    })
})

