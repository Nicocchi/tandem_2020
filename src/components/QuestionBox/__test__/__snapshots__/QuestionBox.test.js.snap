// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`QuestionBox matches snapshop 1`] = `
<div>
  <p
    className="question animated fade-in"
  >
    NaN
     / 
  </p>
  <p
    className="question animated fade-in"
    data-testid="question"
  >
    What is the national animal of Scotland?
  </p>
  <div
    className="answers-box"
    style={
      Object {
        "display": "flex",
      }
    }
  />
  <div
    className="answer-wrapper"
  />
</div>
`;
